core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,299 @@
+        table.io(), table.ops().current().specsById()).entries()) {
+      if (entry.file().path().equals(FILE_A.path())) {
+        Assert.assertEquals("the sequence number of data file should be 1", 1, entry.sequenceNumber().longValue());
+      }
+
+      if (entry.file().path().equals(FILE_B.path())) {
+        Assert.assertEquals("the sequence number of data file should be 2", 2, entry.sequenceNumber().longValue());
+      }
+
+      if (entry.file().path().equals(FILE_C.path()) || entry.file().path().equals(FILE_D.path())) {
+        Assert.assertEquals("the sequence number of data file should be 3", 3, entry.sequenceNumber().longValue());
+      }
+    }
+  }
+
+  @Test
+  public void testSequenceNumberForRewrite() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Assert.assertEquals(2, table.currentSnapshot().sequenceNumber());
+
+    table.rewriteManifests().clusterBy(file -> "").commit();
+    Assert.assertEquals("the sequence number of snapshot should be 3",
+        3, table.currentSnapshot().sequenceNumber());
+
+    ManifestFile newManifest = table.currentSnapshot().manifests().get(0);
+    Assert.assertEquals("the sequence number of manifest should be 3",
+        3, newManifest.sequenceNumber());
+
+    for (ManifestEntry entry : ManifestFiles.read(newManifest,
+        table.io(), table.ops().current().specsById()).entries()) {
+      if (entry.file().path().equals(FILE_A.path())) {
+        Assert.assertEquals("the sequence number of data file should be 1", 1, entry.sequenceNumber().longValue());
+      }
+
+      if (entry.file().path().equals(FILE_B.path())) {
+        Assert.assertEquals("the sequence number of data file should be 1", 2, entry.sequenceNumber().longValue());
+      }
+    }
+  }
+
+  @Test
+  public void testCommitConflict() {
+    Transaction txn = table.newTransaction();
+
+    txn.newFastAppend().appendFile(FILE_A).apply();
+    table.newFastAppend().appendFile(FILE_B).commit();
+
+    AssertHelpers.assertThrows("Should failed due to conflict",
+        IllegalStateException.class, "last operation has not committed", txn::commitTransaction);
+
+    Assert.assertEquals(1, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    AppendFiles appendFiles = table.newFastAppend().appendFile(FILE_C);
+    appendFiles.apply();
+    table.newFastAppend().appendFile(FILE_D).commit();
+    appendFiles.commit();
+
+    ManifestFile manifestFile = table.currentSnapshot().manifests().stream()
+        .filter(manifest -> manifest.snapshotId() == table.currentSnapshot().snapshotId())
+        .collect(Collectors.toList()).get(0);
+
+    for (ManifestEntry entry : ManifestFiles.read(manifestFile, table.io(),
+        table.ops().current().specsById()).entries()) {
+      if (entry.file().path().equals(FILE_C.path())) {
+        Assert.assertEquals(table.currentSnapshot().sequenceNumber(), entry.sequenceNumber().longValue());
```

Comment:
When values are predictable, like sequence numbers, we usually want to test that they are specific values, not that they match some other variable. It would be better to test that FILE_C has sequence number 3.

This also needs to validate that FILE_D was committed first and has sequence number 2.

Reply:
OK